File: backend/app/models/course.py
"""课程、班级、班级成员模型."""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    String,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Course(Base):
    __tablename__ = "courses"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    code: Mapped[str] = mapped_column(String(32), unique=True, nullable=False)
    is_archived: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    classes: Mapped[list[Class]] = relationship(
        back_populates="course", lazy="selectin"
    )


class Class(Base):
    __tablename__ = "classes"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(64), nullable=False)
    course_id: Mapped[int] = mapped_column(ForeignKey("courses.id"), nullable=False)
    teacher_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    student_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    is_archived: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    course: Mapped[Course] = relationship(back_populates="classes", lazy="selectin")
    memberships: Mapped[list[ClassMembership]] = relationship(
        back_populates="class_",
        lazy="selectin",
        cascade="all, delete-orphan",
    )


class ClassMembership(Base):
    """学生加入班级的关系（多对多）."""

    __tablename__ = "class_memberships"
    __table_args__ = (
        UniqueConstraint(
            "class_id", "student_id", name="uq_class_memberships_class_student"
        ),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    class_id: Mapped[int] = mapped_column(
        ForeignKey("classes.id", ondelete="CASCADE"), nullable=False, index=True
    )
    student_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    joined_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    class_: Mapped[Class] = relationship(back_populates="memberships", lazy="selectin")

File: go-backend/internal/model/course.go
package model

import "time"

// Course represents an academic course.
type Course struct {
	ID         int64     `json:"id"`
	Name       string    `json:"name"`
	Code       string    `json:"code"`
	IsArchived bool      `json:"is_archived"`
	CreatedAt  time.Time `json:"created_at"`
	UpdatedAt  time.Time `json:"updated_at"`

	Classes []Class `json:"classes,omitempty"`
}

// Class represents a class within a course.
type Class struct {
	ID           int64     `json:"id"`
	Name         string    `json:"name"`
	CourseID     int64     `json:"course_id"`
	TeacherID    int64     `json:"teacher_id"`
	StudentCount int       `json:"student_count"`
	IsArchived   bool      `json:"is_archived"`
	CreatedAt    time.Time `json:"created_at"`
}

// ClassMembership represents a student's membership in a class.
type ClassMembership struct {
	ID        int64     `json:"id"`
	ClassID   int64     `json:"class_id"`
	StudentID int64     `json:"student_id"`
	JoinedAt  time.Time `json:"joined_at"`
}
